### PR TITLE
Add the `flat` argument to `QueryBuilder.all()` method

### PR DESCRIPTION
**Pull request by sphuber**
This allows to simplify the commonly found form:

    results = [entry[0] for entry in builder.all()]

to the more compact and readable form:

    results = builder.all(flat=True)

The default is set to `flat=False` such that this change is fully
backwards compatible. Occurrences in the code base itself have been
updated as much as possible.

### CODE REVIEW

File: aiida/orm/groups.py [hunk 1]
```diff
@@ -340,21 +340,12 @@ def get(cls, **kwargs):
...
-        for key, val in kwargs.items():
-            filters[key] = val
-
-        query.append(cls, filters=filters)
-        results = query.all()
-        if len(results) > 1:
-            raise exceptions.MultipleObjectsError("Found {} groups matching criteria '{}'".format(len(results), kwargs))
-        if not results:
-            raise exceptions.NotExistent("No group found matching criteria '{}'".format(kwargs))
-        return results[0][0]
```
Comment (ramirezfranciscof) [inline aiida/orm/groups.py:357]: Are these modifications related to the topic? Is it even doing the same thing? At least it would seem that some exceptions would be missed, no?

File: tests/orm/test_querybuilder.py [hunk 2]
```diff
@@ -629,6 +629,32 @@ def test_direction_keyword(self):
         res2 = {item[1] for item in qb.all()}
         self.assertEqual(res2, {d2.id, d4.id})
 
+    @staticmethod
+    def test_flat():
+        """Test the `flat` keyword for the `.iterall()`, `.all()` and `.one()` methods."""
```
Comment (ramirezfranciscof) [inline tests/orm/test_querybuilder.py:634]: Mmm how is this testing `.iterall()` and `.one()`? Is this some copy/paste leftover?

File: aiida/orm/groups.py (hunk 1 above)
Comment (sphuber) [reply aiida/orm/groups.py:357]: It would be, except that if you look at the behavior of `QueryBuilder.one()` already has exactly the same behavior, so this code that I removed was just redundant. The behavior should remain exactly the same. If you think it doesn't, let me know, because then I messed up ;)

File: tests/orm/test_querybuilder.py (hunk 2 above)
Comment (sphuber) [reply tests/orm/test_querybuilder.py:634]: Yeah, this is just a mistake that I left in. I originally thought I was going to implement it for all, but for `one` it would break backwards-compatibility and for `iterall` it does not make sense

File: aiida/orm/groups.py (hunk 1 above)
Comment (ramirezfranciscof) [reply aiida/orm/groups.py:357]: Ah, I see, yes, thats true. The only difference is in the second check `if not results` vs `elif len(res) == 0`, but I can't think of a case where this would cause a divergent behaviour, let alone one which would actually be problematic. I mean, if res/result was a float let's say, it would have problems when evaluating `elif len(res) == 0` but would pass `if not results`, although in that case it would already had problems with the first `if len(results) > 1`. So no, I guess no complaints here.

File: aiida/orm/querybuilder.py [hunk 3]
```diff
@@ -2203,7 +2203,12 @@ def all(self, batch_size=None):
```
Comment (chrisjsewell) [inline aiida/orm/querybuilder.py:2203]:
```suggestion
        :param bool flat: 
            Return the result as a flat list of projected entities,
            with no sub-lists.

```

Comment (chrisjsewell) [reply aiida/orm/querybuilder.py:2203]: Apart from that, it looks good to me 👍
